<a href="https://colab.research.google.com/github/nathlia/Aulas2023/blob/RECUPERA%C3%87%C3%83O-DE-INFORMA%C3%87%C3%83O-E-PERSONALIZA%C3%87%C3%83O-NA-WEB-(DPADP0152)/Entrega1_nathalia.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

import tfidVectorizer


# Imports e iniciando as files

In [ ]:
import numpy as np
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.metrics.pairwise import cosine_similarity

In [ ]:
from google.colab import files

# Use the files.upload() method to upload files
uploaded = files.upload()


dezipar(n usei)

In [ ]:
import zipfile

# Specify the path to the zip file
zip_file_path = '/content/drive/MyDrive/tech.zip'

# Specify the directory where you want to extract the contents
extract_dir = '/content/drive/MyDrive/tech'

# Open the zip file and extract its contents
with zipfile.ZipFile(zip_file_path, 'r') as zip_ref:
    zip_ref.extractall(extract_dir)


importando os stopwords

In [ ]:
## Library imports
import numpy as np
import pandas as pd

from sklearn.feature_extraction.text import CountVectorizer

import os, glob, re, sys, random, unicodedata, collections
from tqdm import tqdm
from functools import reduce
import nltk
nltk.download('stopwords')
from collections import Counter

from nltk.corpus import stopwords
from nltk.stem import RSLPStemmer
from nltk.tokenize import sent_tokenize , word_tokenize

STOP_WORDS = set(stopwords.words('english'))

In [48]:
print(STOP_WORDS)

{'once', 'off', 'when', 'and', 'with', 'are', 'into', 'below', 'herself', 'after', 'then', 'doing', "won't", "you'll", 'had', "she's", 'a', 'does', 'his', 'isn', 'ours', 'there', "hadn't", 'up', 'each', 'ma', 'until', 'what', 'being', 'not', 'same', 'over', 'hadn', 'before', 'whom', 's', "shan't", 'it', "needn't", 'am', 'at', "haven't", 'my', 'as', 'hers', 'their', 'should', 'needn', 'won', 'o', "you've", 'your', "doesn't", 'her', "you'd", 'about', 'has', 'how', 'themselves', "couldn't", 'both', 'other', 'some', "isn't", 'by', "wouldn't", 'if', 'these', 'on', 'doesn', "mustn't", 'did', 'any', 'more', 'wouldn', 'didn', 'myself', 'they', 'during', 'such', 'while', "aren't", 'because', 'can', 'of', 'couldn', 'yourself', 'but', 'wasn', "it's", "wasn't", 'where', 'll', 're', 'out', 'shan', 'weren', "hasn't", 'you', 't', 'further', 'have', 'been', 'from', 'aren', 'that', 'theirs', "that'll", 'which', 'ourselves', 'again', "didn't", 'me', 'were', 'for', 'to', 've', "shouldn't", 'i', 'than', '

In [55]:
stop_words_list = list(STOP_WORDS)
print(stop_words_list)


['once', 'off', 'when', 'and', 'with', 'are', 'into', 'below', 'herself', 'after', 'then', 'doing', "won't", "you'll", 'had', "she's", 'a', 'does', 'his', 'isn', 'ours', 'there', "hadn't", 'up', 'each', 'ma', 'until', 'what', 'being', 'not', 'same', 'over', 'hadn', 'before', 'whom', 's', "shan't", 'it', "needn't", 'am', 'at', "haven't", 'my', 'as', 'hers', 'their', 'should', 'needn', 'won', 'o', "you've", 'your', "doesn't", 'her', "you'd", 'about', 'has', 'how', 'themselves', "couldn't", 'both', 'other', 'some', "isn't", 'by', "wouldn't", 'if', 'these', 'on', 'doesn', "mustn't", 'did', 'any', 'more', 'wouldn', 'didn', 'myself', 'they', 'during', 'such', 'while', "aren't", 'because', 'can', 'of', 'couldn', 'yourself', 'but', 'wasn', "it's", "wasn't", 'where', 'll', 're', 'out', 'shan', 'weren', "hasn't", 'you', 't', 'further', 'have', 'been', 'from', 'aren', 'that', 'theirs', "that'll", 'which', 'ourselves', 'again', "didn't", 'me', 'were', 'for', 'to', 've', "shouldn't", 'i', 'than', '

In [263]:
# count term frequency using CountVectorizer from scikit-learn
## limiting number of words just for illustrating the concept

vec = CountVectorizer(max_features=20, stop_words=stop_words_list)
X = vec.fit_transform(docs)
df = pd.DataFrame(X.toarray(), columns=vec.get_feature_names_out())
files_names = [file.split('/')[-1] for file in sample_texts]
df['file'] = files_names
df = df.set_index('file')

print(df)

         also  could  digital  games  many  mobile  mr  music  net  new  one  \
file                                                                           
023.txt     2      0        0      0     0       0   3      0    0    0    0   
020.txt     2      0        0      0     1       0   1      0    0    6    1   
018.txt     2      3        1      3     0       0   0      0    0    1    0   
016.txt     0      2        0      0     1       0   5      0    0    1    6   
021.txt     0      1        0      0     2       0   1      0    1    1    1   
...       ...    ...      ...    ...   ...     ...  ..    ...  ...  ...  ...   
004.txt     1      2        4      1     0       1   4      0    0    0    3   
003.txt     1      0        0      0     0       0   0      0    0    0    0   
002.txt     3      0        0      1     1       0   0      0   15    1    1   
401.txt     0      5        0     22     4       0   0      0    0    2    7   
001.txt     0      0        0      0    

mounted google drive para acessar os textos, escolhi a pasta tech e importei para o drive

In [7]:
from google.colab import drive
drive.mount('/content/drive/MyDrive')

Mounted at /content/drive/MyDrive


In [11]:
import os

# Path to your Google Drive directory
google_drive_path = '/content/drive/MyDrive/MyDrive/ENTREGA1/tech'

# List the files and subdirectories in your Google Drive directory
file_list = os.listdir(google_drive_path)

# Count the number of files
file_count = len(file_list)

# Print the count
print(f"Number of files in your Google Drive directory: {file_count}")

Number of files in your Google Drive directory: 401


In [26]:
import random


In [30]:
cwd = os.getcwd()  # Get the current working directory (cwd)
files = os.listdir(google_drive_path)  # Get all the files in that directory
print("Files in %r: %s" % (cwd, files))

Files in '/content': ['023.txt', '020.txt', '018.txt', '016.txt', '021.txt', '015.txt', '017.txt', '019.txt', '022.txt', '014.txt', '012.txt', '010.txt', '008.txt', '013.txt', '009.txt', '005.txt', '007.txt', '011.txt', '006.txt', '024.txt', '054.txt', '053.txt', '050.txt', '045.txt', '049.txt', '047.txt', '048.txt', '046.txt', '051.txt', '052.txt', '044.txt', '033.txt', '026.txt', '031.txt', '028.txt', '029.txt', '025.txt', '027.txt', '043.txt', '037.txt', '041.txt', '040.txt', '035.txt', '036.txt', '038.txt', '039.txt', '042.txt', '034.txt', '030.txt', '032.txt', '084.txt', '073.txt', '066.txt', '065.txt', '068.txt', '069.txt', '071.txt', '083.txt', '080.txt', '078.txt', '081.txt', '075.txt', '076.txt', '079.txt', '077.txt', '082.txt', '074.txt', '067.txt', '070.txt', '063.txt', '060.txt', '059.txt', '055.txt', '061.txt', '057.txt', '058.txt', '072.txt', '064.txt', '056.txt', '062.txt', '113.txt', '111.txt', '110.txt', '106.txt', '105.txt', '107.txt', '109.txt', '108.txt', '112.txt',

In [198]:
# select and read 10 random files
sample_texts = files
# sample_texts = random.sample(file_list, 100)
print("Lista de textos:\n", sample_texts)

docs = []
for fname in sample_texts:
    with open(google_drive_path + '/' + fname , "r", encoding="ISO-8859-1") as file:
        text = file.read()
    docs.append(text)

print("Finished reading files in:", google_drive_path)

Lista de textos:
 ['023.txt', '020.txt', '018.txt', '016.txt', '021.txt', '015.txt', '017.txt', '019.txt', '022.txt', '014.txt', '012.txt', '010.txt', '008.txt', '013.txt', '009.txt', '005.txt', '007.txt', '011.txt', '006.txt', '024.txt', '054.txt', '053.txt', '050.txt', '045.txt', '049.txt', '047.txt', '048.txt', '046.txt', '051.txt', '052.txt', '044.txt', '033.txt', '026.txt', '031.txt', '028.txt', '029.txt', '025.txt', '027.txt', '043.txt', '037.txt', '041.txt', '040.txt', '035.txt', '036.txt', '038.txt', '039.txt', '042.txt', '034.txt', '030.txt', '032.txt', '084.txt', '073.txt', '066.txt', '065.txt', '068.txt', '069.txt', '071.txt', '083.txt', '080.txt', '078.txt', '081.txt', '075.txt', '076.txt', '079.txt', '077.txt', '082.txt', '074.txt', '067.txt', '070.txt', '063.txt', '060.txt', '059.txt', '055.txt', '061.txt', '057.txt', '058.txt', '072.txt', '064.txt', '056.txt', '062.txt', '113.txt', '111.txt', '110.txt', '106.txt', '105.txt', '107.txt', '109.txt', '108.txt', '112.txt', '1

# Consultas

Elabore 6 exemplos de consulta.

In [291]:
# Define queries
queries = [
    "Document security and information leakage in Microsoft Word",
    "Consumer Electronics Show (CES) 2005 highlights",
    "Digital games and technology",
    "Impact of technology on people",
    "Web links and virus",
    "Cyber-crime",
]

for q in queries:
    print(q)

Document security and information leakage in Microsoft Word
Consumer Electronics Show (CES) 2005 highlights
Digital games and technology
Impact of technology on people
Web links and virus
Cyber-crime


In [87]:
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.metrics.pairwise import cosine_similarity

Primeira rodada: sem retirar stopword e sem "stemming"



# Utils

Calcula Similiariedade do Coseno

In [233]:
def calcule_cosine_similarities(query_matrix, document_matrix):
    cosine = cosine_similarity(query_matrix, document_matrix)
    return cosine

Print resultado da busca

Considera o documento como valido dependendo do valor da similiariedade de coseno, ordena em ordem decrescente o valor de similiaredade dos documentos validos.  

In [301]:
# Calcular a média das similaridades para cada consulta
similaridade_media = [sum(sim_scores) / len(sim_scores) for sim_scores in cosine_similarities]


# Print matching document titles and file names based on similarity threshold
def print_similarity(queries, cosine_similarities, documents, file_names, threshold):
    highest_similarity = -1.0  # Initialize with a value lower than any possible similarity
    lowest_similarity = 2.0  # Initialize with a value higher than any possible similarity
    highest_similarity_info = None
    lowest_similarity_info = None

    for i, query in enumerate(queries):
        print('---' * 30)
        print(f"⭐ Query {i + 1}: {query} ⭐")
        print(f"Similaridade Média: {similaridade_media[i]:.4f}")  # Adicione esta linha
        print('---' * 30)
        similar_documents = []

        for j, sim_score in enumerate(cosine_similarities[i]):
            if sim_score > threshold:  # Adjust the similarity threshold as needed
                document_title = documents[j].split('\n')[0]  # Extract the first line (title)
                similar_documents.append((document_title, file_names[j], sim_score))

                # Update highest and lowest similarity scores and their corresponding info
                if sim_score > highest_similarity:
                    highest_similarity = sim_score
                    highest_similarity_info = (document_title, file_names[j], sim_score)
                if sim_score < lowest_similarity:
                    lowest_similarity = sim_score
                    lowest_similarity_info = (document_title, file_names[j], sim_score)

        if similar_documents:
            num = 0
            similar_documents.sort(key=lambda x: x[2], reverse=True)
            for title, file_name, score in similar_documents:
                num = num + 1;
                print(f"📄 {num} File: {file_name} ∽ Similarity: {score:.4f} ✒️Title: {title}")
        else:
            print("No matching titles found.")
        print("\n")

    print(f"Highest Similarity: {highest_similarity:.4f}")
    if highest_similarity_info:
        title, file_name, score = highest_similarity_info
        print(f"Highest Similarity Title: {title}")
        print(f"Highest Similarity File Name: {file_name}")
        print(f"Highest Similarity Score: {score:.4f}")

    print(f"Lowest Similarity: {lowest_similarity:.4f}")
    if lowest_similarity_info:
        title, file_name, score = lowest_similarity_info
        print(f"Lowest Similarity Title: {title}")
        print(f"Lowest Similarity File Name: {file_name}")
        print(f"Lowest Similarity Score: {score:.4f}")

# Primeira rodada: sem retirar stopword e sem "stemming"


Cria um objeto TfidfVectorizer chamado de **vectorizer** que converte os dados do text em um vetor **TF-IDF** (Term Frequency-Inverse Document Frequency), que é uma **representação numérica de documentos de texto**.

A lista de documentos é transformada em uma matriz usando o metodo de fit_trasform do vectorizer.

Calcula similiariedade de cosenos entre os vetores das queries e os vetores dos documentos.

In [302]:
def find_matching_titles(queries, documents, file_names):
    # Create a TfidfVectorizer
    vectorizer = TfidfVectorizer()

    # Fit and transform the documents to obtain TF-IDF vectors
    tfidf_matrix = vectorizer.fit_transform(documents)

    # Transform the queries into TF-IDF vectors using the same vectorizer
    query_tfidf_matrix = vectorizer.transform(queries)

    # Calculate cosine similarities between queries and documents
    cosine_similarities = calcule_cosine_similarities(query_tfidf_matrix, tfidf_matrix)

    print_similarity(queries, cosine_similarities, documents, file_names, 0)

# Example usage
find_matching_titles(queries, docs, file_list)

------------------------------------------------------------------------------------------
⭐ Query 1: Document security and information leakage in Microsoft Word ⭐
Similaridade Média: 0.0315
------------------------------------------------------------------------------------------
📄 1 File: 086.txt ∽ Similarity: 0.4431 ✒️Title: Warning over Windows Word files
📄 2 File: 083.txt ∽ Similarity: 0.1926 ✒️Title: Microsoft makes anti-piracy move
📄 3 File: 308.txt ∽ Similarity: 0.1926 ✒️Title: Microsoft makes anti-piracy move
📄 4 File: 036.txt ∽ Similarity: 0.1833 ✒️Title: Microsoft seeking spyware trojan
📄 5 File: 003.txt ∽ Similarity: 0.1833 ✒️Title: Microsoft seeking spyware trojan
📄 6 File: 369.txt ∽ Similarity: 0.1766 ✒️Title: Microsoft plans 'safer ID' system
📄 7 File: 060.txt ∽ Similarity: 0.1643 ✒️Title: Microsoft releases patches
📄 8 File: 007.txt ∽ Similarity: 0.1639 ✒️Title: Microsoft releases bumper patches
📄 9 File: 020.txt ∽ Similarity: 0.1587 ✒️Title: Security scares spark brows

## Comentário:

Possui numero variaveis de similaridade entre os documentos, o que afirma que os documentos possuem um nivel diferente de relevancia entre as buscas.

A maior similaridade encontrada foi na Busca 1 "Document security and information leakage in Microsoft Word", sendo ela 0.4431 do documento "Warning over Windows Word files", que foi feita especificamente para esse texto.

A unicas buscas que não mostraram todos os documentos foram:
*   Busca 2: "Consumer Electronics Show (CES) 2005 highlights"
*   Busca 6: "Cyber-crime"

Que foram retiradas diretamente do texto.

Coloquei uma condicao de só mostrar documentos com similaridade maior que 0. Mesmo assim 4 das 6 buscam possuem todos os documentos como similares.

Comparando com o resultado do codigo exemplo do ChatGPT, nota-se que esse método possui resultados de similiriaedade maiores que o metodo do ChatGPT, mas continua mostrando a mesma quantidade de documentos.


### Exemplo ChatGPT da aula

In [294]:
# Initialize the TfidfVectorizer
vectorizer = TfidfVectorizer()

# Fit and transform your documents and queries
tfidf_matrix = vectorizer.fit_transform(docs + queries)

# Calculate the cosine similarities between the queries and documents
cosine_similarities = cosine_similarity(tfidf_matrix[-len(queries):], tfidf_matrix[:-len(queries)])


# Print only the titles of matching documents for each query
for i, query in enumerate(queries):
    sorted_indices = cosine_similarities[i].argsort()[::-1]
    print('---' * 20)
    print(f"☆ Query {i + 1}: {query} ☆")
    print('---' * 20)
    print("\nMatching Document Titles:")
    for idx in sorted_indices:
        document_title = docs[idx].split('\n')[0]  # Extract the first line (title)
        if cosine_similarities[i][idx] >  0:
            print(f"Similarity: {cosine_similarities[i][idx]:.4f} - Title: {document_title}")
    print('\n')

------------------------------------------------------------
☆ Query 1: Document security and information leakage in Microsoft Word ☆
------------------------------------------------------------

Matching Document Titles:
Similarity: 0.3399 - Title: Warning over Windows Word files
Similarity: 0.1503 - Title: Microsoft makes anti-piracy move
Similarity: 0.1503 - Title: Microsoft makes anti-piracy move
Similarity: 0.1432 - Title: Microsoft seeking spyware trojan
Similarity: 0.1432 - Title: Microsoft seeking spyware trojan
Similarity: 0.1380 - Title: Microsoft plans 'safer ID' system
Similarity: 0.1282 - Title: Microsoft releases patches
Similarity: 0.1279 - Title: Microsoft releases bumper patches
Similarity: 0.1239 - Title: Security scares spark browser fix
Similarity: 0.1191 - Title: 'Blog' picked as word of the year
Similarity: 0.1181 - Title: Microsoft debuts security tools
Similarity: 0.1181 - Title: Microsoft debuts security tools
Similarity: 0.1164 - Title: Solutions to net securi

#  Segunda rodada: retirando stopwords e sem "stemming"




## Stopwords


In [303]:
def find_matching_titles_segunda(queries, documents, file_names):
    # Create a TfidfVectorizer with stopwords removed
    vectorizer = TfidfVectorizer(stop_words=stop_words_list)

    # Fit and transform the documents to obtain TF-IDF vectors
    tfidf_matrix = vectorizer.fit_transform(documents)

    # Transform the queries into TF-IDF vectors using the same vectorizer
    query_tfidf_matrix = vectorizer.transform(queries)

    # Calculate cosine similarities between queries and documents
    cosine_similarities = calcule_cosine_similarities(query_tfidf_matrix, tfidf_matrix)

    print_similarity(queries, cosine_similarities, documents, file_names, 0)

# Example usage
find_matching_titles_segunda(queries, docs, file_list)

------------------------------------------------------------------------------------------
⭐ Query 1: Document security and information leakage in Microsoft Word ⭐
Similaridade Média: 0.0315
------------------------------------------------------------------------------------------
📄 1 File: 086.txt ∽ Similarity: 0.4693 ✒️Title: Warning over Windows Word files
📄 2 File: 083.txt ∽ Similarity: 0.1919 ✒️Title: Microsoft makes anti-piracy move
📄 3 File: 308.txt ∽ Similarity: 0.1919 ✒️Title: Microsoft makes anti-piracy move
📄 4 File: 036.txt ∽ Similarity: 0.1716 ✒️Title: Microsoft seeking spyware trojan
📄 5 File: 003.txt ∽ Similarity: 0.1716 ✒️Title: Microsoft seeking spyware trojan
📄 6 File: 369.txt ∽ Similarity: 0.1685 ✒️Title: Microsoft plans 'safer ID' system
📄 7 File: 060.txt ∽ Similarity: 0.1601 ✒️Title: Microsoft releases patches
📄 8 File: 007.txt ∽ Similarity: 0.1597 ✒️Title: Microsoft releases bumper patches
📄 9 File: 020.txt ∽ Similarity: 0.1540 ✒️Title: Security scares spark brows

### Comentário

Com a remoção das stopwords, nota-se que nenhuma das buscas retornam todos os documentos, e todas estão com um número total diferente de documentos. A quantidade de resultados diminuiu consideravelmente.

O valor da similaridade dos primeiros itens amentaram levemente também. Mas no caso da busca 4, ela caiu.

Maior similaridade continua a mesma, enquando a menor mudou.


# Terceira rodada: retirando stopwords e  com "stemming"

# Conceitos

### Bag of Words (BoW)

The Bag of Words approach is like making a list of all the unique words you find in these books, without caring about the order or how many times each word appears. It's like counting how many times each word shows up but not worrying about the order in which they appear.

In [154]:
BoW = vectorizer.fit_transform(file_list)

print(BoW)

  (0, 401)	0.15668736734446556
  (0, 22)	0.9876482516132251
  (1, 19)	0.9876482516132251
  (1, 401)	0.15668736734446556
  (2, 17)	0.9876482516132251
  (2, 401)	0.15668736734446556
  (3, 15)	0.9876482516132251
  (3, 401)	0.15668736734446556
  (4, 20)	0.9876482516132251
  (4, 401)	0.15668736734446556
  (5, 14)	0.9876482516132251
  (5, 401)	0.15668736734446556
  (6, 16)	0.9876482516132251
  (6, 401)	0.15668736734446556
  (7, 18)	0.9876482516132251
  (7, 401)	0.15668736734446556
  (8, 21)	0.9876482516132251
  (8, 401)	0.15668736734446556
  (9, 13)	0.9876482516132251
  (9, 401)	0.15668736734446556
  (10, 11)	0.9876482516132251
  (10, 401)	0.15668736734446556
  (11, 9)	0.9876482516132251
  (11, 401)	0.15668736734446556
  (12, 7)	0.9876482516132251
  :	:
  (388, 401)	0.15668736734446556
  (389, 371)	0.9876482516132251
  (389, 401)	0.15668736734446556
  (390, 399)	0.9876482516132251
  (390, 401)	0.15668736734446556
  (391, 394)	0.9876482516132251
  (391, 401)	0.15668736734446556
  (392, 398)	0

### N-grams

This modification will create a matrix where each column represents a bigram (a combination of two words).

In [147]:
n_gramsVectorizer = CountVectorizer(ngram_range=(2, 2))

n_grams = n_gramsVectorizer.fit_transform(file_list)

print(n_grams)

  (0, 22)	1
  (1, 19)	1
  (2, 17)	1
  (3, 15)	1
  (4, 20)	1
  (5, 14)	1
  (6, 16)	1
  (7, 18)	1
  (8, 21)	1
  (9, 13)	1
  (10, 11)	1
  (11, 9)	1
  (12, 7)	1
  (13, 12)	1
  (14, 8)	1
  (15, 4)	1
  (16, 6)	1
  (17, 10)	1
  (18, 5)	1
  (19, 23)	1
  (20, 53)	1
  (21, 52)	1
  (22, 49)	1
  (23, 44)	1
  (24, 48)	1
  :	:
  (376, 376)	1
  (377, 379)	1
  (378, 380)	1
  (379, 381)	1
  (380, 372)	1
  (381, 364)	1
  (382, 370)	1
  (383, 369)	1
  (384, 367)	1
  (385, 368)	1
  (386, 373)	1
  (387, 365)	1
  (388, 366)	1
  (389, 371)	1
  (390, 399)	1
  (391, 394)	1
  (392, 398)	1
  (393, 395)	1
  (394, 397)	1
  (395, 396)	1
  (396, 3)	1
  (397, 2)	1
  (398, 1)	1
  (399, 400)	1
  (400, 0)	1


### HashingVectorizer
The n_features parameter determines the number of features in the output matrix. HashingVectorizer doesn't build a vocabulary like CountVectorizer.

In [148]:
from sklearn.feature_extraction.text import HashingVectorizer

hashingV = HashingVectorizer(n_features=1000)

hashing = hashingV.transform(file_list)

print(hashing)

  (0, 551)	-0.7071067811865475
  (0, 869)	-0.7071067811865475
  (1, 662)	0.7071067811865475
  (1, 869)	-0.7071067811865475
  (2, 761)	-0.7071067811865475
  (2, 869)	-0.7071067811865475
  (3, 493)	0.7071067811865475
  (3, 869)	-0.7071067811865475
  (4, 716)	-0.7071067811865475
  (4, 869)	-0.7071067811865475
  (5, 832)	-0.7071067811865475
  (5, 869)	-0.7071067811865475
  (6, 402)	-0.7071067811865475
  (6, 869)	-0.7071067811865475
  (7, 15)	-0.7071067811865475
  (7, 869)	-0.7071067811865475
  (8, 241)	0.7071067811865475
  (8, 869)	-0.7071067811865475
  (9, 717)	0.7071067811865475
  (9, 869)	-0.7071067811865475
  (10, 12)	-0.7071067811865475
  (10, 869)	-0.7071067811865475
  (11, 399)	0.7071067811865475
  (11, 869)	-0.7071067811865475
  (12, 537)	0.7071067811865475
  :	:
  (388, 869)	-0.7071067811865475
  (389, 111)	0.7071067811865475
  (389, 869)	-0.7071067811865475
  (390, 34)	-0.7071067811865475
  (390, 869)	-0.7071067811865475
  (391, 310)	0.7071067811865475
  (391, 869)	-0.70710678118

### Word vs. Character N-grams

You can specify whether you want word or character N-grams using the analyzer parameter. For example, to work with character bigrams (N=2) with word boundaries (like " j" for "jump"), you can do this:

This will create a matrix where each column represents a character bigram with word boundaries.*italicized text*

In [152]:
bigrams_with_boundariesV = CountVectorizer(analyzer='char_wb', ngram_range=(2, 2))

bigrams = bigrams_with_boundariesV.fit_transform(file_list)

print(bigrams)

  (0, 0)	1
  (0, 9)	1
  (0, 32)	1
  (0, 39)	1
  (0, 5)	1
  (0, 117)	1
  (0, 118)	1
  (0, 116)	1
  (1, 0)	1
  (1, 9)	1
  (1, 5)	1
  (1, 117)	1
  (1, 118)	1
  (1, 116)	1
  (1, 29)	1
  (1, 6)	1
  (2, 0)	1
  (2, 5)	1
  (2, 117)	1
  (2, 118)	1
  (2, 116)	1
  (2, 8)	1
  (2, 26)	1
  (2, 94)	1
  (3, 0)	1
  :	:
  (397, 10)	1
  (398, 0)	1
  (398, 9)	1
  (398, 5)	1
  (398, 117)	1
  (398, 118)	1
  (398, 116)	1
  (398, 28)	1
  (398, 7)	1
  (399, 5)	1
  (399, 117)	1
  (399, 118)	1
  (399, 116)	1
  (399, 8)	1
  (399, 17)	1
  (399, 51)	1
  (399, 4)	1
  (400, 0)	1
  (400, 5)	1
  (400, 117)	1
  (400, 118)	1
  (400, 116)	1
  (400, 8)	1
  (400, 17)	1
  (400, 7)	1
